LoRaEngine/modules/LoRaWanNetworkSrvModule/Logger/LogSink.cs

```diff
@@ -0,0 +1,295 @@
+            {
+                SingleReader = true,
+                FullMode = BoundedChannelFullMode.DropOldest,
+            });
+            this.logger = logger;
+        }
+
+        public void Dispose()
+        {
+            try
+            {
+                this.cancellationTokenSource.Cancel();
+            }
+            catch (ObjectDisposedException)
+            {
+                return; // bail out because this means object has already been disposed before
+            }
+            catch (AggregateException ex)
+            {
+                // "AggregateException" is thrown and contains all the exceptions thrown by the
+                // registered callbacks on the associated "CancellationToken". However,
+                // "IDisposable.Dispose" methods are not expected to throw exceptions thus it is
+                // simply ignore here.
+
+                Debug.WriteLine(ex);
+            }
+
+            this.cancellationTokenSource.Dispose();
+        }
+
+        private async Task SendAllLogMessagesAsync(CancellationToken cancellationToken)
+        {
+            var encoding = Encoding.UTF8;
+            var buffers = ArrayPool<byte>.Shared;
+            var client = new TcpClient();
+
+            try
+            {
+                await foreach (var log in this.channel.Reader.ReadAllAsync(cancellationToken))
+                {
+                    var lines = log.IndexOfAny(NewLineChars) >= 0
+                              ? new StringValues(SplitIntoLines(log).Append(string.Empty).ToArray()) // re-normalize
+                              : new StringValues(log);
+
+                    for (var attempt = 1; ; attempt++)
+                    {
+                        foreach (var line in lines)
+                        {
+                            try
+                            {
+                                if (!client.Connected)
+                                {
+                                    this.logger?.LogDebug($"Connecting to log server at (attempt {attempt}/{this.maxRetryAttempts}): " + this.serverEndpoint);
+                                    await client.ConnectAsync(this.serverEndpoint.Address, this.serverEndpoint.Port);
+                                }
+
+                                var size = encoding.GetByteCount(line) + 2;
+                                var buffer = buffers.Rent(size);
+                                var bi = encoding.GetBytes(line, buffer);
+                                buffer[bi++] = 13; // CR
+                                buffer[bi++] = 10; // LF
+                                Debug.Assert(bi == size);
+
+                                try
+                                {
+                                    await client.GetStream().WriteAsync(buffer.AsMemory(..bi), cancellationToken);
+                                    break; // don't retry on success
```

Comment:
I think there's a bug in this `break` statement. With that we break out of the inner loop, but not of the outer loop. As a result we will send the log statement many times

Reply:
Very good catch. I refactored this a bit late with c61b03b0281667f2e925b43753b37634c3850b5d and inverted the loops but this is now fixed with 06f0af3e945ad25543b32a567b34da48aa63f6f9.